[FEAT/#33] SharePhoto 화면전환 시 필요한 Entity를 구성하였습니다.

hsw1920 opened the pull request:
## 🤔 배경
- `EditPhotoRoomFeature`에서 `SharePhotoRoomFeature`로 화면전환 시 담아 보낼 데이터 구조가 필요했습니다.
- 해당 데이터는 현재 `ImageData` 하나는 확정이며 추가될 수 있기 때문에 `SharePhotoComponent`로 모아서 전달할 수 있도록 `DomainInterface`에 `Entity`를 두었습니다.
- `FrameImageGenerator`의 `frameView` 인스턴스가 존재하지 않는 문제가 있었습니다.

## 📃 작업 내역
- `SharePhotoComponent` 구현
- `SharePhotoViewModel`의 `imageData`를 `SharePhotoViewController`가 참조하여 자신의 `photoView`에 표현하는 기능 구현
- `FrameImageGenerator` 버그 픽스

## ✅ 리뷰 노트
- 이벤트 스트림을 처리할 필요가 아직 없기 때문에 리액티브는 구현하지 않았습니다.
- `FrameImageGenerator` 버그 픽스의 경우 아래 트러블 슈팅을 작성하였습니다.
  -  [didSet 트러블 슈팅](https://www.notion.so/0hooni/didSet-dfadfb2173d547229e2b42d9e4429a7b?pvs=4)

## 🎨 스크린샷
| SharePhotoFeature | 공유 시트 | 파일 생성 | 
| -------------- |  ---- |  --- | 
| <img width="264" alt="image" src="https://github.com/user-attachments/assets/e305806e-a448-4a4f-8635-d88d967980c8"> | <img width="257" alt="image" src="https://github.com/user-attachments/assets/0cf3b0a9-87af-49d8-baf5-2f6afdec58e0"> | <img width="265" alt="image" src="https://github.com/user-attachments/assets/eeb7efa0-cde8-4a7d-9f86-182a443e5ad3"> |


## 🚀 테스트 방법
`SharePhotoFeatureDemo`를 실행해주세요.

hsw1920 committed:
fix/#33 :: 초기화 시점에 frameView 인스턴스 생성

- didSet에서 frameView 생성이 안되는 문제를 해결

hsw1920 committed:
feat/#33 :: SharePhoto 컴포넌트 구현

- EditPhotoRoomFeature에서 SharePhotoRoomFeature로 화면전환될 때 필요한 의존성을 구현하였습니다.
- 현재는 image를 Data 타입으로 담아 보내며, 필요한 의존성이 추가될 수 있습니다.

hsw1920 committed:
feat/#33 :: SharePhotoViewModel 구현

- SharePhotoViewModel은 SharePhotoViewController에 필요한 모델을 가집니다.
- 현재 imageData의 경우 리액티브가 필요 없으므로 viewModel에서 참조하여 자신의 PhotoImageView에 넣어줍니다.

hsw1920 committed:
feat/#33 :: SharePhotoFeatureDemo 구성 변경

FrameImageGenerator로부터 이미지를 생성하고, 해당 이미지를 Data 타입으로 변환하여 SharePhotoComponent에 담아 의존성을 주입하는 과정을 표현합니다.

hsw1920 committed:
chore/#33 :: update pbxproj

Kiyoung-Kim-57 commented on PhotoGether/PresentationLayer/SharePhotoFeature/SharePhotoFeature/Source/SharePhotoViewModel.swift:
```diff
@@ -0,0 +1,11 @@
+import PhotoGetherDomainInterface
+
+public final class SharePhotoViewModel {
+    public private(set) var photoData: Data
```
ViewModel에서 import UIKit 없이 데이터로 저장해서 사용하는게 좋네요!

youn9k commented on PhotoGether/DomainLayer/PhotoGetherDomain/PhotoGetherDomainInterface/Entity/SharePhotoComponent.swift:
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+public struct SharePhotoComponent {
```
Component 라는 네이밍이 디자인 시스템에서 사용하는 Component와 겹쳐 뜻이 모호하게 느껴져요.
디렉토리도 Entity에 둔만큼 SharePhotoEntity가 더 자연스러운 것 같아요!

youn9k replied:
Share라는 접두사가 붙어 해당 모델이 어디에 사용될 지 알고 있는 느낌이에요
PhotoEntity는 어떤가요?

hsw1920 replied:
오히려 PhotoEntity가 직접적으로 관계가 없으며, **SharePhoto에 필요한 의존성**을 생각하고 네이밍을 했습니다.
해당 모델이 어디에 사용될 지 알고 있다는 표현이 맞습니다.
그렇다고 SharePhotoFeatureInterface를 만들기에는 과하니깐 DomainInterface에 두었습니다.

현재는 필요한 의존성이 image만 존재하기때문에 PhotoEntity가 적합해보이는 오해를 불러일으킨 것 같습니다.

RIBs에서 RIB을 구성하기 위해서 필요한 의존성들을 Component라고 칭하는 것을 생각하여 Component를 차용했습니다.

<img width="645" alt="image" src="https://github.com/user-attachments/assets/bd541924-91f7-498d-8bb5-f50126fc43e4">

youn9k replied:
EditPhotoFeature <->  SharePhotoFeature 간에 소통하기 위한 모델이라면 이해가 되는 것 같습니다.
그러면 Domain 레이어와 Data 레이어간 소통하기 위한 모델인 Entity와 Presentation 레이어에서 사용하는 모델인 Component를 분리하기 위해 `/Component` 디렉토리를 추가하면 어떨까요?

hsw1920 replied:
레이어간 소통모델에 대한 분리를 위해 DomainInterface에서 `/Entity`와 `/Component`로 나누어 관리하는 것이 좋아보이네요!

[chore/](https://github.com/boostcampwm-2024/iOS04-PhotoGether/pull/38/commits/96ffb894fe5c6187c07b35b6eaa8a51effa9b32d)https://github.com/boostcampwm-2024/iOS04-PhotoGether/issues/33 [:: Rename Directory](https://github.com/boostcampwm-2024/iOS04-PhotoGether/pull/38/commits/96ffb894fe5c6187c07b35b6eaa8a51effa9b32d)에서 수정 완료했습니다.

디렉터리를 `/Entity`에서 `/Component`로 변경하였습니다.

꼼꼼한 리뷰 감사합니다~

Kiyoung-Kim-57 reviewed (approved):
엘쥐티엠 👁️o👁️

youn9k commented on PhotoGether/DomainLayer/PhotoGetherDomain/PhotoGetherDomainInterface/Entity/SharePhotoComponent.swift:
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+public struct SharePhotoComponent {
+    public let image: Data
```
image 라는 변수명이 조금 추상적인 것 같아요. 사용하는 쪽에서 아래와 같이 사용될 때 코드가 모호해질 수 있을 것 같아요.
Data 타입임을 유추할 수 있도록 `imageData` 는 어떨까요?
```swift
let myImage: UIImage = UIImage(data: image)

```

hsw1920 replied:
[45ec9c0](https://github.com/boostcampwm-2024/iOS04-PhotoGether/pull/38/commits/45ec9c08ef7d033552d7d0294e3ccbe90c4fdd5c) 에서 해당 코멘트 반영하였습니다.

youn9k commented on PhotoGether/PresentationLayer/SharePhotoFeature/SharePhotoFeature/Source/SharePhotoViewController.swift:
```diff
@@ -84,8 +92,7 @@ public class SharePhotoViewController: BaseViewController, ViewControllerConfigu
-    // TODO: 권한 요청 및 사진앱에 저장
-    private func saveImage() {
-        
+    private func convertToImage(with data: Data) -> UIImage? {
+        return UIImage(data: data)
```
bypass만 하고 있는 것 같아요. 추가적인 로직이 들어갈 여지가 있어서 함수를 만들어 두신걸까요?

hsw1920 replied:
저는 파라미터에서 인스턴스를 생성하는것을 선호하지 않아서
```swift
        photoView.image = UIImage(data: viewModel.photoData)
```
가 아니라,
```swift
        let imageData = viewModel.photoData
        let image = UIImage(data: imageData)
        photoView.image = image
```
이런식으로 명시적으로 3줄을 쓰곤 하는데요,

이러면 `configureUI()`가 지저분해서 따로 뺐습니당

개인적으로 전자는 가독성 측면에서 좋지 않다고 생각합니다.

`let viewController = ViewController(viewModel:ViewModel(useCase: UseCase(repository: Repository(...))))))` 이런걸 피하고 있습니당

youn9k commented on PhotoGether/PresentationLayer/SharePhotoFeature/SharePhotoFeatureDemo/App/SceneDelegate.swift:
```diff
@@ -1,5 +1,8 @@
 import UIKit
 import SharePhotoFeature
+import EditPhotoRoomFeature
+import PhotoGetherDomainInterface
+import DesignSystem
 
```
컨벤션 너무 맛있어요 우마이

hsw1920 replied:
넵 [style/#33 :: 라이브러리 import 컨벤션 수정](https://github.com/boostcampwm-2024/iOS04-PhotoGether/pull/38/commits/f7b729f6cd5e26fb38083e51b0eeaaf5f23991f2) 에서 수정완료!

hsw1920 committed:
style/#33 :: 프로퍼티 네이밍 변경

- Data 타입을 유추하기 쉽도록 프로퍼티명을 변경하였습니다.

hsw1920 committed:
style/#33 :: 라이브러리 import 컨벤션 수정

- A-Z 순서로 변경

youn9k reviewed (approved):
고생하셨습니다!!

hsw1920 committed:
chore/#33 :: Rename Directory

- DataLayer와 DomainLayer가 소통하는 Entity와 Presentation과 Domain이 소통하기위한 Component를 분리하기 위한 디렉터리 네이밍을 변경
